### ISSUE TITLE
Bug - mutations with optimisticUpdate are failing since 2.9.1 

### ISSUE THREAD

**Issue by roomman** [zenstackhq/zenstack#1878]
@ymc9 I've updated to 2.9.1 which includes [your PR #1842](https://github.com/zenstackhq/zenstack/pull/1842), and I am seeing failing mutations, and no network requests being made, when `optimisticUpdate: true`. 

The cases that are failing include optional "belongs-to" relationships where no relationship has yet been formed. For example:

```
model SomeModel {
  id                           String   @id @default(uuid())
  title                        String
  business               Business?          @relation(fields: [businessId], references: [id], onDelete: Cascade)
  businessId            String?

}
```
I am seeing this error:
`Error - there was an error updating SomeModel TypeError: Cannot convert undefined or null to object`

The error is about to the "business" relationship from the above schema, and an evaluation on this line from mutator.ts: https://github.com/zenstackhq/zenstack/blob/a003be9ee46daf791550993cf2b595f4ed9f9214/packages/runtime/src/cross/mutator.ts#L165

Because no business has been related to SomeModel the value of both `business` and `businessId` is `null`. Your change expects either an array or an object. 

If I set `optimisticUpdates: false` for the mutation the error goes away and mutations are successful.

Should L165 be doing a null check?

**Comment by roomman** [comment]
@ymc9 I've created a [minimal reproduction](https://github.com/roomman/zenstack-minimal-repro-uptimistic-update-bug).

**Set up**:
- Clone the repo
- Run `npm install`
- Run `npm run generate`
- Run `npm run seed`
- Run `npm run dev`

Open your browser and you will see that you can click on an "Enter Competition" button for any Competition, including those with no related Business and it works.

Next run `npm run up` to install the latest and greatest. Run `npm run dev` again and retry entering a competition. Check the dev tools and see failed mutations. Under state your should see the following error: `{"name":"TypeError","message":"Cannot convert undefined or null to object"}`

Got to src/app/page.tsx L37 and comment out the optimistic update:

```tsx
  // create a new competition entry
  const { mutate: enterCompetition } = useCreateCompetitionEntry({
    // optimisticUpdate: true,
  });
```

Refresh and retry entering a competition and note that the mutations are now working as expected.

Revert the change to L37. Instead, comment out L14-19 (i.e. don't include the optional relationship):

```tsx
  // fetch all competitions
  const { data: allCompetitions, isPending } = useFindManyCompetition({
    include: {
      // business: {
      //   select: {
      //     id: true,
      //     name: true,
      //   },
      // },
      entries: true,
    },
  });
```

Refresh and retry entering a competition and note the mutations are working still.

As I suggested in my original post, the inclusion of an optional relationship, where the default value returned by Prisma is `null`, is not accounted for in the evaluation on L165 of mutator.ts.

**Comment by ymc9** [comment]
Hi @roomman , thanks for reporting the issue. Let me have a check and follow up.

**Comment by ymc9** [comment]
Yes, you're right, a null checking is missing there. Probably safer to check it to be a non-null and object. Will be interested in creating a PR for it 😄 @roomman

**Comment by ymc9** [comment]
> Yes, @ymc9, happy to. Will sort asap. 👍🏼

Thank you. Appreciate it!

**Comment by roomman** [comment]
Yes, was working on the tests this afternoon so hopefully that something tomorrow morning.

**Comment by roomman** [comment]
@ymc9 just a quick sense check if you have the time?

If I fetch all Users and include deeply nested relationships, like this:

```
useModelQuery(
    'User',
    makeUrl('User', 'findMany'),
    {
        include: {
            posts: {
                include: {
                    category: true,
                },
            },
        },
    },
    { optimisticUpdate: true }
),
```
And then added a new Post and "connect" it to a Category (where a Post can have one Category, and a Category can have many Posts), would you expect the User in the cache to have:

```
// where the category id and nested category object is included
{
    id: "1",
    name: "user1",
    posts: [
        {
            id: "1",
            title: "post1",
            categoryId: "1",
            category: {
                id: "1",
                name: "category1"
            }
        }
    ]
}
```
Or: 
```
// where the nested category object is committed:
{
    id: "1",
    name: "user1",
    posts: [
        {
            id: "1",
            title: "post1",
            categoryId: "1"
        }
    ]
}
```
I had been testing on the basis that the Visitor would add the object too, but I only get the foreign key. Not had time to get into the internals of the Visitor and mindful that, with other users registering the issue, we need a quick fix.

Perhaps I should fix the null issue now and then skip the nested object test any look at that later, to you can take that up if it's better?

**Comment by roomman** [comment]
@ymc9 I've submitted [the PR](https://github.com/zenstackhq/zenstack/pull/1884) with a couple of notes. Cheers 👍🏼

**Comment by ymc9** [comment]
Thank you! The change looks great to me. I've merged it and will publish a new patch release.

**Comment by ymc9** [comment]
Fixed in 2.9.3
